### DISCUSSION TITLE
LoadFS

### DISCUSSION THREAD

**Discussion by khrome83** [pocketbase/pocketbase#3440] [Q&A]
For templates, how I kind of expected it to work, would be to use the `embed` directive. 

Where I load the views directory. 

```go
//go:embed views/**/*.gohtml
var fs embed.FS
```

Then later when I register the files, I would do so the normal way, but pass a glob that would match the directive, or have no need to pass a glob at all. 

I know I can use `registry.cache.Get("key")` to grab a renderer, or even call `LoadFS()` again. 

But Ideally it seems like when I define the embed FS in the Registry, I should just be able to call Render, and pass the paths that are needed, or better yet, just the entry point.

Right now I have to pass `fs` into the handler, so i can maintain the key structure, and still call `LoadFS()`, which seems strange if I am embedding the whole folder.

I started looking into this, because I had everything working, then found out I didn't copy this file to the docker image. I assumed maybe something was being embedded under the hood. 

Using embed would also remove the need for the cache and key layer, which could simplify usage.

**Comment by ganigeorgiev** [comment]
I'm not sure I follow.

Could you elaborate with more concrete example and/or provide a pseudo-code with the expected behavior?

**Reply by khrome83** [reply]
Sure, sorry little under the weather, so not giving my best effort.


### Main File
```go
// IDEAL STATE
//go:embed views/**/*.gohtml
var fs embed.FS

func main() {
  app := pocketbase.New()
  registry := template.NewRegistry()
 
 // NOT IDEAL STATE - does not really do anything, 
 // because the files I need are specific to the rendered page in the handler, so would always be a miss due to cache
  registry.LoadFS(fs, "views/**/*.gohtml")

 // IDEAL STATE - it does not need a glob since embed.FS is only what was in the glob in the directive
  registry.LoadEmbedFS(fs)

  app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
     e.Router.GET("/hello", handlers.GetHello(app))
  })

  if err := app.Start(); err != nil {
	log.Fatal(err)
  }
}
```

### Handler
```go
package handlers

func GetHello(app *pocketbase.PocketBase, registry *template.Registry) echo.HandlerFunc {
    return func(c echo.Context) error {

        // OMIT CODE, including using *pocketbase.PocketBase app

        // Option 1 - Load Files will go to Disk, so can't use Embed
		html, err := registry.LoadFiles(
			"views/layouts/main.gohtml",
			"views/hello.gohtml",
		).Render(map[string]any{
			"Title":     "Hello",
		})

        // Option 2 - Load FS - Now I have to pass FS around, in addition to PocketBase and Registry
        // This is because you use fs memory space value as part of the key
        // And have a cache, because your not counting on embed existing. 
               html, err := registry.LoadFS(fs, "views/layouts/main.gohtml",
			"views/hello.gohtml",
		  ).Render(map[string]any{
			  "Title":     "Hello",
		  })

         // IDEAL STATE - Assume Load Embed FS where registry was created (or config on registry creation)
         // No need to pass FS around. It is already embeded and attached. 
         // Cache is not needed, so no need for key creation and management as embed puts everything in the app itself.
               html, err := registry.UseEmbedFS("views/layouts/main.gohtml",
			"views/hello.gohtml").Render(map[string]any{
			  "Title":     "Hello",
		  })

		if err != nil {
			// or redirect to a dedicated 404 HTML page
			return apis.NewNotFoundError("", err)
		}

		return c.HTML(http.StatusOK, html)
    }    
})
```

I used Embed FS here to keep backwards compatibility with the older style. In a more perfect world, if I had 3 partials, ideally, `hello.gohtml` rendering would not need me to specify the 3 partials because the reference to them is already in the embed. So you just need entry point.

Fiber does something like this, except you can specify a optional layout that wraps the pages entry point (injects at `{{embed}}` If i remember right).

**Reply by ganigeorgiev** [reply]
I think I understand what you mean but I'm not sure if an extra abstraction/helper layer is needed for this case.

`LoadFS` is similar to `LoadFiles`, meaning that you don't have to call `LoadFS` in the main file. The main difference between those 2 methods is that `LoadFS` uses the provided `fs.FS` interface as source for the template files instead of the local filesystem:

```go
html, err := registry.LoadFiles(
    "views/layouts/main.gohtml",
    "views/hello.gohtml",
).Render(nil)

html, err := registry.LoadFS(
    fs,
    "views/layouts/main.gohtml",
    "views/hello.gohtml",
).Render(nil)
```

The cache key of the `LoadFS` templates would be the string representation of the `fs` + the filenames/glob patterns (for `LoadFS` is just the filenames).

In both cases the template parsing happens only once, on the first render.

Also note that you can load more than one `fs` in the same template registry, aka. we can't use `UseEmbedFS` or similar because it would be ambiguous.

If the whole problem comes from the fact that you have pass an extra `fs` argument then one possible approach similar to the linked fiber docs could be to create a small helper struct like:

```go
type EmbedRegistry struct {
	reg *template.Registry
	fs  fs.FS
}

func (r *EmbedRegistry) Load(patterns ...string) *template.Renderer {
	return r.reg.LoadFS(r.fs, patterns...)
}
```

```go
// you can pass an instance of the above struct to your handler
embedRegistry := EmbedRegistry{registry, fs}

...

// and then in the handler you can call
embedRenderer.Load("base.html", "hello.html").Render(data)
```


Depending on the users demand I guess we could add a helper method that will create the above `EmbedRegistry` struct but it could cause confusion with the `LoadFS` method and for now I think it is better without it.

**Reply by khrome83** [reply]
It's not exactly what I am getting at. I this case, you have to have the files on disk. 

If you did embed directive they would be in the binary, and there would not need a need to transfer the files separately to a docker image or maintain a cache as it's already in memory and not reading from disk.

In theory this would also simplify migrations as you would not need to transfer those either, they could be baked into the build. 

I discovered when moving these renders to docker, that I had a missing file error. And its because the .gohtml are not using embed directive. 

It's just a suggestion to make it cleaner. Technical with the directive you can have as many as you want. Because this was specific to templates I didn't see the need to specify which one. 

But the difference is going to disk/having to ensure the files make the trip to docker or deploy, vs everything in a single binary. 

It's super minor, I just think it's one of the benefits of golang that they make it so easy to use.

Out of curiosity, why is App and Registry not attached to the echo context? Or rather would it be a benefit to have Pocketbase have its own context with those attached? Would be less on the user to pass around.

**Reply by ganigeorgiev** [reply]
> ... And its because the .gohtml are not using embed directive.

I'm confused and I don't understand what do you mean by this. When you have the `embed` directive and you build the binary, then  they will be part of it, this is not PocketBase specific.

If you think that this can be simplified, please elaborate or file a PR, otherwise I think we are going in circles.

**Reply by khrome83** [reply]
Sorry @ganigeorgiev I don't mean to cause confusion or go around in circles. 

You are correct that the embed directive is not specific to PocketBase. I think all I am saying is I can't use LoadFiles with the directive, if my preference was to have the files embedded. I have to use LoadFS. The downside of this is I now have another thing to pass around the handlers because of the cache layer and key creation.

Since `app *pocketbase.PocketBase`, `registry *template.Registry`, and `fs fs*FS` are not on the echo context, every handler I build is now a custom wrapper to export a handler. It's a lot of boilerplate. I was looking to reduce that a little further.

Once option i see, is PocketBase has a version of the echo Context that includes common things if they were configured (didn't show a example of this), and always adds app. 

Another option, specifically just to `fs`, is using the `embed` directive, which removes the need for you managing a cache and passing it between handlers, because the `fs` point is no longer used in the key. I think you could namespace this to allow multiple `fs` which is something you support now.

It's really just about making the interfaces cleaner and reducing code. And making it one less set of external files that need to be deployed with the app.

None of it is needed. My goal is not to be critical. It's just syntactical sugar I have fallen in love with from other frameworks. But I think would make it easier to use and document to users.

Do you have any interest in a PR for this? I can over the next few weeks try and find some time to build out the PocketBase context wrapper, or the embed directive to better illustrate. Which path would you be more open too?

Would you also be open to a PR to talk about static files, and views being transferred in Docker, probably inserted here?

```Dockerfile
# uncomment to copy the local pb_migrations dir into the image
# COPY ./pb_migrations /pb/pb_migrations

# uncomment to copy the local pb_hooks dir into the image
# COPY ./pb_hooks /pb/pb_hooks
```

Sorry for all the confusion. You have been a immense help, and I am blown away with what is built so far. And i appreciate how stable and thoughtful the project is.

**Reply by ganigeorgiev** [reply]
Your comments always contains a lot of text but I'm really failing to understand what actually you mean and refer to (_I'm not a native speaker so sorry for this_).

There are no plans to add extra stuff to the `echo.Context` as it will be unnecessary in my opinion and you'll lose the typesafety this way. There are plans to refactor the routing and even maybe replace echo entirely with the [upcoming router enhancements in the standard library](https://github.com/golang/go/issues/61410) but for now is pushed aside and other things are with higher priority.

Again, PR for improvements or to simplify the process are welcomed. There are no promises that it will get merged but at least I can have a look at it and finally understand what you are trying to do.

**Reply by khrome83** [reply]
That gives me some direction. Sorry @ganigeorgiev I will work on how to convey things better for you. 

Great to know about the router enhancements. I personally have been having issues with Echo v5, because it's hard to find documentation and updated middleware. Using a middleware v4 results in a echo specific error says echo v4 is not compatibility with echo v5. Understand it will be a while, if at all.

**Reply by ganigeorgiev** [reply]
And just to clear some eventual misunderstanding, if you use the `embed` directive you don't have to transfer anything in your Dockerfile, because the files will be part of the resulting binary.

You need to copy the files that the `embed` directive is pointing in a Dockerfile only if you are building your application in a Docker, but this is the same for the regular source files.

**Comment by khrome83** [comment]
https://github.com/gofiber/template/blob/master/html/README.md

As a example of a interface that is a little cleaner.
